### ISSUE TITLE
Integration test does not wait before closing HTTP Servers

### ISSUE THREAD

**Issue by klarose** [envoyproxy/go-control-plane#114]
I want to make some changes to go-control-plane locally. Prior to making any modifications, I decided to run the integration tests before doing anything else. Surprisingly, they failed:
```
make integration.docker 
--> building Linux test binary
--> building test docker image
Sending build context to Docker daemon  182.3MB
Step 1/5 : FROM envoyproxy/envoy:latest
 ---> cf2442a79328
Step 2/5 : ADD sample /sample
 ---> Using cache
 ---> 901328cc36a6
Step 3/5 : ADD build/integration.sh build/integration.sh
 ---> Using cache
 ---> e9f5d2931ab0
Step 4/5 : ADD bin/test-linux /bin/test
 ---> Using cache
 ---> 21abea4c5a7c
Step 5/5 : ENTRYPOINT build/integration.sh
 ---> Using cache
 ---> c430b556d974
Successfully built c430b556d974
Successfully tagged test:latest
docker run -it -e "XDS=ads" test -debug
INFO[0000] upstream listening HTTP/1.1                   port=18080
INFO[0000] waiting for the first request...             
INFO[0000] gateway listening HTTP/1.1                    port=18001
INFO[0000] access log server listening                   port=18090
INFO[0000] management server listening                   port=18000
ERRO[0000] http: Server closed                          
ERRO[0000] http: Server closed                          
DEBU[0000] stream 1 open for                            
DEBU[0000] open watch 1 for type.googleapis.com/envoy.api.v2.Cluster[] from nodeID "test-id", version "" 
INFO[0000] initial snapshot {Xds:ads Version: UpstreamPort:18080 BasePort:9000 NumClusters:4 NumHTTPListeners:2 NumTCPListeners:2} 
INFO[0000] executing sequence                            requests=5 updates=3
INFO[0000] update snapshot                               version=v0
DEBU[0000] respond open watch 1[] with new version "v0" 
DEBU[0000] respond type.googleapis.com/envoy.api.v2.Cluster[] version "" with version "v0" 
DEBU[0000] respond type.googleapis.com/envoy.api.v2.ClusterLoadAssignment[cluster-v0-3 cluster-v0-2 cluster-v0-1 cluster-v0-0] version "" with version "v0" 
DEBU[0000] open watch 2 for type.googleapis.com/envoy.api.v2.Cluster[] from nodeID "test-id", version "v0" 
INFO[0000] request batch                                 batch=0 failed=4 ok=0 pass=false
DEBU[0000] respond type.googleapis.com/envoy.api.v2.Listener[] version "" with version "v0" 
DEBU[0000] open watch 3 for type.googleapis.com/envoy.api.v2.ClusterLoadAssignment[cluster-v0-3 cluster-v0-2 cluster-v0-1 cluster-v0-0] from nodeID "test-id", version "v0" 
DEBU[0000] respond type.googleapis.com/envoy.api.v2.RouteConfiguration[route-v0-1 route-v0-0] version "" with version "v0" 
DEBU[0000] open watch 4 for type.googleapis.com/envoy.api.v2.Listener[] from nodeID "test-id", version "v0" 
DEBU[0000] open watch 5 for type.googleapis.com/envoy.api.v2.RouteConfiguration[route-v0-1 route-v0-0] from nodeID "test-id", version "v0" 
INFO[0000] request batch                                 batch=1 failed=4 ok=0 pass=false
INFO[0001] request batch                                 batch=2 failed=4 ok=0 pass=false
INFO[0001] request batch                                 batch=3 failed=4 ok=0 pass=false
INFO[0002] request batch                                 batch=4 failed=4 ok=0 pass=false
INFO[0002] server callbacks                              fetches=0 requests=8
ERRO[0002] failed all requests in a run 0               
Envoy log: envoy.log
Makefile:89: recipe for target 'integration.docker' failed
make: *** [integration.docker] Error 1
```
Note these lines:

> ERRO[0000] http: Server closed                          
> ERRO[0000] http: Server closed  

If you look at the code for one, you can see:

```go
func RunHTTP(ctx context.Context, upstreamPort uint) {
	log.WithFields(log.Fields{"port": upstreamPort}).Info("upstream listening HTTP/1.1")
	server := &http.Server{Addr: fmt.Sprintf(":%d", upstreamPort), Handler: echo{}}
	go func() {
		if err := server.ListenAndServe(); err != nil {
			log.Error(err)
		}
	}()
	if err := server.Shutdown(ctx); err != nil {
		log.Error(err)
	}
}
```
It's not waiting for func() to actually finish, so it just closes the server immediately. I'm guessing that on my laptop, this happens almost always, whereas in circleci the timing is different?

I've fixed it locally by using ctx.Done() like the gRPC server methods.

I'm using go 1.11.
I tested this against e79e039 and revert_grpc_version_update

I'll submit a PR soon.

**Comment by klarose** [comment]
PR: https://github.com/envoyproxy/go-control-plane/pull/115

**Comment by klarose** [comment]
Closed by #115
